File: java/src/main/java/com/microsoft/microhack/catalog/service/CatalogDocumentParser.java
package com.microsoft.microhack.catalog.service;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.type.CollectionType;
import com.microsoft.microhack.catalog.model.CatalogImportItem;
import jakarta.validation.ConstraintViolation;
import jakarta.validation.Validator;
import java.io.IOException;
import java.io.InputStream;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.UUID;
import java.util.regex.Pattern;
import org.springframework.stereotype.Component;

/** Parses and validates complete catalog documents without publishing partial prefixes. */
@Component
public class CatalogDocumentParser {

    private static final Pattern CANONICAL_UUID = Pattern.compile(
            "^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$");
    private final ObjectMapper objectMapper;
    private final Validator validator;

    public CatalogDocumentParser(ObjectMapper objectMapper, Validator validator) {
        this.objectMapper = objectMapper.copy()
                .enable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES);
        this.validator = validator;
    }

    /** Returns only after every record and cross-record identity has been validated. */
    public List<ValidatedCatalogItem> parse(InputStream input) {
        List<CatalogImportItem> records = readDocument(input);
        if (records.isEmpty()) {
            throw new CatalogImportValidationException("catalog document must contain at least one record");
        }
        Map<String, String> namesToSlugs = new HashMap<>();
        Map<String, String> slugsToNames = new HashMap<>();
        return records.stream()
                .map(record -> validate(record, namesToSlugs, slugsToNames))
                .toList();
    }

    private List<CatalogImportItem> readDocument(InputStream input) {
        try {
            CollectionType listType = objectMapper.getTypeFactory()
                    .constructCollectionType(List.class, CatalogImportItem.class);
            return objectMapper.readValue(input, listType);
        } catch (JsonProcessingException exception) {
            throw new CatalogImportValidationException("catalog document is not valid JSON", exception);
        } catch (IOException exception) {
            throw new CatalogImportValidationException("catalog document could not be read", exception);
        }
    }

    private ValidatedCatalogItem validate(
            CatalogImportItem record,
            Map<String, String> namesToSlugs,
            Map<String, String> slugsToNames) {
        Set<ConstraintViolation<CatalogImportItem>> violations = validator.validate(record);
        if (!violations.isEmpty()) {
            throw new CatalogImportValidationException(
                    "catalog record violates " + violations.iterator().next().getPropertyPath());
        }
        if (!CANONICAL_UUID.matcher(record.productId()).matches()) {
            throw new CatalogImportValidationException(
                    "productId must be a canonical lowercase UUID");
        }
        UUID productId;
        try {
            productId = UUID.fromString(record.productId());
        } catch (IllegalArgumentException exception) {
            throw new CatalogImportValidationException("productId is not an RFC 4122 UUID", exception);
        }
        if (!record.productId().equals(productId.toString())) {
            throw new CatalogImportValidationException(
                    "productId must be a canonical lowercase UUID");
        }
        String expectedFilename = record.productId() + ".png";
        if (!expectedFilename.equals(record.filename())) {
            throw new CatalogImportValidationException(
                    "filename must equal productId plus .png");
        }
        String slug = CategorySlug.normalize(record.category());
        if (slug.isEmpty() || slug.length() > 64) {
            throw new CatalogImportValidationException(
                    "category must normalize to a non-empty slug of at most 64 characters");
        }
        String existingSlug = namesToSlugs.putIfAbsent(record.category(), slug);
        String existingName = slugsToNames.putIfAbsent(slug, record.category());
        if ((existingSlug != null && !existingSlug.equals(slug))
                || (existingName != null && !existingName.equals(record.category()))) {
            throw new CatalogImportValidationException(
                    "category display names and slugs must be unique");
        }
        return new ValidatedCatalogItem(
                productId,
                record.name(),
                record.description(),
                record.category(),
                slug,
                record.filename());
    }
}

File: dotnet/src/LegoCatalog.App/Services/CatalogDocumentParser.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace LegoCatalog.App.Services;

/// <summary>
/// Parses and validates a complete catalog document before publication starts.
/// </summary>
public sealed class CatalogDocumentParser
{
    private static readonly JsonSerializerOptions SerializerOptions = new()
    {
        PropertyNameCaseInsensitive = false,
        UnmappedMemberHandling = JsonUnmappedMemberHandling.Disallow,
    };

    /// <summary>
    /// Returns fully validated records or rejects the entire document.
    /// </summary>
    public async Task<IReadOnlyList<ValidatedCatalogItem>> ParseAsync(
        Stream jsonStream,
        CancellationToken cancellationToken)
    {
        List<CatalogInputItem>? rawItems;
        try
        {
            rawItems = await JsonSerializer.DeserializeAsync<List<CatalogInputItem>>(
                jsonStream,
                SerializerOptions,
                cancellationToken);
        }
        catch (JsonException exception)
        {
            throw new CatalogImportValidationException(
                "The catalog document is not valid contract JSON.",
                1,
                exception);
        }

        if (rawItems is null || rawItems.Count == 0)
        {
            throw new CatalogImportValidationException(
                "The catalog document must contain at least one record.",
                1);
        }

        var validated = new List<ValidatedCatalogItem>(rawItems.Count);
        var ids = new HashSet<Guid>();
        for (var index = 0; index < rawItems.Count; index++)
        {
            try
            {
                validated.Add(Validate(rawItems[index], ids));
            }
            catch (CatalogImportValidationException exception)
            {
                throw new CatalogImportValidationException(
                    $"Record {index + 1} is invalid: {exception.Message}",
                    rawItems.Count,
                    exception);
            }
        }

        return validated;
    }

    private static ValidatedCatalogItem Validate(
        CatalogInputItem item,
        ISet<Guid> ids)
    {
        RequireLength(item.Name, 3, 80, "name");
        RequireLength(item.Description, 20, 1200, "description");
        RequireLength(item.Category, 2, 60, "category");
        RequireLength(item.ImagePrompt, 30, 260, "imagePrompt");
        if (item.ProductId is null
            || !Guid.TryParseExact(item.ProductId, "D", out var productId)
            || !string.Equals(
                productId.ToString("D"),
                item.ProductId,
                StringComparison.Ordinal))
        {
            throw new CatalogImportValidationException(
                "productId must be a canonical lowercase UUID.",
                1);
        }

        if (!ids.Add(productId))
        {
            throw new CatalogImportValidationException(
                "productId values must be unique within one document.",
                1);
        }

        var expectedFilename = $"{productId:D}.png";
        if (!string.Equals(
            item.Filename,
            expectedFilename,
            StringComparison.Ordinal))
        {
            throw new CatalogImportValidationException(
                $"filename must equal {expectedFilename}.",
                1);
        }

        var category = item.Category!;
        var categorySlug = CategorySlug.Normalize(category);
        if (categorySlug.Length == 0)
        {
            throw new CatalogImportValidationException(
                "category must normalize to a non-empty slug.",
                1);
        }

        return new ValidatedCatalogItem(
            productId,
            item.Name!,
            item.Description!,
            category,
            categorySlug,
            item.Filename!);
    }

    private static void RequireLength(
        string? value,
        int minimum,
        int maximum,
        string field)
    {
        var length = value?.EnumerateRunes().Count() ?? 0;
        if (length < minimum || length > maximum)
        {
            throw new CatalogImportValidationException(
                $"{field} must contain from {minimum} through {maximum} characters.",
                1);
        }
    }

    private sealed record CatalogInputItem(
        [property: JsonPropertyName("productId")] string? ProductId,
        [property: JsonPropertyName("name")] string? Name,
        [property: JsonPropertyName("description")] string? Description,
        [property: JsonPropertyName("category")] string? Category,
        [property: JsonPropertyName("filename")] string? Filename,
        [property: JsonPropertyName("imagePrompt")] string? ImagePrompt);
}

public sealed record ValidatedCatalogItem(
    Guid ProductId,
    string Name,
    string Description,
    string Category,
    string CategorySlug,
    string Filename);

/// <summary>
/// Signals whole-document validation failure before publication.
/// </summary>
public sealed class CatalogImportValidationException : Exception
{
    public CatalogImportValidationException(
        string message,
        int rejectedCount,
        Exception? innerException = null)
        : base(message, innerException)
    {
        RejectedCount = rejectedCount;
    }

    public int RejectedCount { get; }
}
